speechbrain/k2_integration/lattice_decoder.py

```diff
@@ -0,0 +1,339 @@
+    )
+
+    if compose_HL_with_G or use_G_rescoring:
+        logger.info("Converting arpa LM(s) to FST(s)")
+        G_path = Path(hparams["lm_dir"]) / hparams["G_fst_output_name"]
+        G_rescoring_path = (
+            Path(hparams["lm_dir"]) / hparams["G_rescoring_fst_output_name"]
+        ) if use_G_rescoring else None
+        lm_dir = Path(hparams["lm_dir"])
+        run_on_main(
+            utils.arpa_to_fst,
+            kwargs={
+                "words_txt": Path(hparams["lang_dir"]) / "words.txt",
+                "in_arpa_files": [ lm_dir / hparams["G_arpa_name"],
+                ] +([lm_dir / hparams["G_rescoring_arpa_name"]] if use_G_rescoring else []),
+                "out_fst_files": [G_path, ] + ([ G_rescoring_path] if use_G_rescoring else []),
+                "lms_ngram_orders": [ 3 ] + ([4] if use_G_rescoring else [])
+            },
+        )
+        assert G_path.is_file(), f"{G_path} does not exist"
+
+    if compose_HL_with_G:
+        logger.info(f"Loading G LM: {G_path}")
+        G = utils.load_G(G_path)
+        decoding_graph = graphCompiler.compile_HLG(G, cache_to=hparams["output_folder"])
+    else:
+        decoding_graph = graphCompiler.compile_HL()
+
+    if not isinstance(hparams["rescoring_lm_scale"], list):
+        hparams["rescoring_lm_scale"] = [hparams["rescoring_lm_scale"]]
+
+    def decoding_method(lattice:k2.Fsa) -> List[str]:
+        raise NotImplementedError(f"{hparams.get('decoding_method')} not implemented as a decoding_method")
+
+    if hparams.get("decoding_method") == "whole-lattice-rescoring":
+        G_rescoring = None
+        def decoding_method(lattice:k2.Fsa) -> k2.Fsa:
+            """Get the best path from a lattice given rescoring_lm_scale."""
+
+            # Lazy load rescoring G (takes a lot of time) for developer happiness
+            nonlocal G_rescoring
+            if G_rescoring is None:
+                logger.info(f"Decoding method: whole-lattice-rescoring")
+                logger.info(f"Loading rescoring LM: {G_rescoring_path}")
+                G_rescoring_pt = utils.load_G(G_rescoring_path)
+                graphCompiler.lexicon.remove_G_rescoring_disambig_symbols(G_rescoring_pt)
+                G_rescoring = utils.prepare_rescoring_G(G_rescoring_pt)
+
+            # rescore_with_whole_lattice returns a list of paths depending on
+            # lm_scale values.
+            return rescore_with_whole_lattice(
+                lattice,
+                G_rescoring,
+                lm_scale_list=hparams["rescoring_lm_scale"],
+            )
+
+    if hparams.get("decoding_method") in ["1best", "onebest"]:
+        logger.info(f"Decoding method: one-best-decoding")
+        def decoding_method(lattice:k2.Fsa) -> k2.Fsa:
+            """Get the best path from a lattice."""
+            return OrderedDict({"1best": one_best_decoding(lattice)})
+
+    return {"decoding_graph":decoding_graph,
+            "decoding_method":decoding_method}
+
+
+torch.no_grad()
```

Comment:
Do you think we should put it here, or in the decoding script?

Reply:
I really think it is better here.
I you want to add other `decoding_method` you can easily do so and make them available for all other egs.